### PR TITLE
Fix DomainsService crash when account.wordPressComRestApi nil

### LINKED ISSUE #22171: EXC_BREAKPOINT: wordPressComRestApi
First Seen 4 days ago in release [23.7.1.0](https://a8c.sentry.io/releases/23.7.1.0/?project=5716771)
Similar issue: https://github.com/wordpress-mobile/WordPress-iOS/pull/20559

### Sentry

Sentry Issue: [JETPACK-IOS-16FY](https://a8c.sentry.io/issues/4693405243/?referrer=github_integration)

```
EXC_BREAKPOINT: wordPressComRestApi
  File "<compiler-generated>", in value
  File "DomainsService.swift", line 229, in DomainsService.init
  File "<compiler-generated>", line 228, in DomainsService.init
  File "AllDomainsListViewModel.swift", line 52, in AllDomainsListViewModel.init
  File "AllDomainsListViewModel.swift", in AllDomainsListViewModel.init
...
(26 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by staskus**
Fixes #22171

In rare cases, `account.wordPressComRestApi` can be `nil`, and causes a crash in `AllDomainsListViewModel`. 
Similar issue: https://github.com/wordpress-mobile/WordPress-iOS/pull/20559

## Solution

Other service calls check for `wordPressComRestApi` since it can be `nil`. The long-term solution would be to make `wordPressComRestApi` nullable, which would force to handle `nil` scenarios in all the cases properly.

Made one of the `DomainsService` initializers failable when initializing with `account.wordPressComRestApi`. 

`AllDomainsListViewModel`already expects optional `DomainsService`, so no change is needed there.

## To test:

The easiest way to reproduce the issue is to return `nil` from `WPAccount.m` `wordPressComRestApi` (line 164). 

### Case 1:
1. Return `nil` from `WPAccount.m` `wordPressComRestApi` (line 164)
2. Launch Jetpack app
3. Log into .com account
4. Me -> All Domains
5. There shouldn't be any crash

## Regression Notes
1. Potential unintended areas of impact

None, in rare cases when `wordPressComRestApi` is `nil`, a sign-in screen should be presented.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/Domains/DomainsService.swift [hunk 1]
```diff
@@ -225,8 +225,9 @@ struct DomainsService {
 }
 
 extension DomainsService {
-    init(coreDataStack: CoreDataStack, account: WPAccount) {
-        self.init(coreDataStack: coreDataStack, remote: DomainsServiceRemote(wordPressComRestApi: account.wordPressComRestApi))
+    init?(coreDataStack: CoreDataStack, account: WPAccount) {
+        guard let wordPressComRestApi = account.wordPressComRestApi else { return nil }
```
Comment (alpavanoglu) [inline WordPress/Classes/Services/Domains/DomainsService.swift]: Why aren't we simply passing in a `wordPressComRestApi` if that's the requirement for the Service to be initialized?

**Review by alpavanoglu — COMMENTED** [review]
✅ Test Case Succeeds

File: WordPress/Classes/Services/Domains/DomainsService.swift (hunk 1 above)
Comment (staskus) [reply WordPress/Classes/Services/Domains/DomainsService.swift]: We can. It was actually how I initially did it but then reverted to limited number of changes. However, I do agree it's a better idea.

**Review by alpavanoglu — APPROVED** [review]
LGTM 🚀

**Pull request merged**